module/zcommon/zfs_fletcher.c

```diff
@@ -222,30 +264,219 @@ fletcher_4_incremental_native(const void *buf, uint64_t size,
+	if (fletcher_4_impl_chosen != idx)
+		fletcher_4_impl_chosen = idx;
+	mutex_exit(&fletcher_4_impl_lock);
+
+	return (0);
+}
+
+static inline const fletcher_4_ops_t *
+fletcher_4_impl_get(void)
+{
+#if !defined(_KERNEL)
+	if (fletcher_4_impl_chosen == FLETCHER_CYCLE) {
+		static volatile unsigned int cycle_count = 0;
+		const fletcher_4_ops_t *ops = NULL;
+		unsigned int index;
+
+		while (1) {
+			index = atomic_inc_uint_nv(&cycle_count);
+			ops = fletcher_4_algos[
+			    index % ARRAY_SIZE(fletcher_4_algos)];
+			if (ops->valid == NULL || ops->valid())
+				break;
+		}
+		return (ops);
+	}
+#endif
+	membar_producer();
+	return (fletcher_4_impl_selectors[fletcher_4_impl_chosen].fis_ops);
+}
+
+void
+fletcher_4_native(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	const fletcher_4_ops_t *ops = fletcher_4_impl_get();
+
+	ops->init(zcp);
+	ops->compute(buf, size, zcp);
+	if (ops->fini != NULL)
+		ops->fini(zcp);
+}
+
+void
+fletcher_4_byteswap(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	const fletcher_4_ops_t *ops = fletcher_4_impl_get();
+
+	ops->init(zcp);
+	ops->compute_byteswap(buf, size, zcp);
+	if (ops->fini != NULL)
+		ops->fini(zcp);
+}
+
+void
+fletcher_4_incremental_native(const void *buf, uint64_t size,
+    zio_cksum_t *zcp)
+{
+	fletcher_4_scalar(buf, size, zcp);
+}
+
 void
 fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
     zio_cksum_t *zcp)
 {
-	const uint32_t *ip = buf;
-	const uint32_t *ipend = ip + (size / sizeof (uint32_t));
-	uint64_t a, b, c, d;
+	fletcher_4_scalar_byteswap(buf, size, zcp);
+}
 
-	a = zcp->zc_word[0];
-	b = zcp->zc_word[1];
-	c = zcp->zc_word[2];
-	d = zcp->zc_word[3];
+void
+fletcher_4_init(void)
+{
+	const uint64_t const bench_ns = (50 * MICROSEC); /* 50ms */
+	unsigned long best_run_count = 0;
+	unsigned long best_run_index = 0;
+	const unsigned data_size = 4096;
+	char *databuf;
+	int i;
 
-	for (; ip < ipend; ip++) {
-		a += BSWAP_32(ip[0]);
-		b += a;
-		c += b;
-		d += c;
+	databuf = kmem_alloc(data_size, KM_SLEEP);
+	for (i = 0; i < ARRAY_SIZE(fletcher_4_algos); i++) {
+		const fletcher_4_ops_t *ops = fletcher_4_algos[i];
+		kstat_named_t *stat = &fletcher_4_kstat_data[i];
+		unsigned long run_count = 0;
+		hrtime_t start;
+		zio_cksum_t zc;
+
+		strncpy(stat->name, ops->name, sizeof (stat->name) - 1);
+		stat->data_type = KSTAT_DATA_UINT64;
+		stat->value.ui64 = 0;
+
+		if (ops->valid != NULL && !ops->valid())
+			continue;
+
+		kpreempt_disable();
+		start = gethrtime();
+		ops->init(&zc);
+		do {
+			ops->compute(databuf, data_size, &zc);
+			run_count++;
+		} while (gethrtime() < start + bench_ns);
+		if (ops->fini != NULL)
+			ops->fini(&zc);
+		kpreempt_enable();
+
+		if (run_count > best_run_count) {
+			best_run_count = run_count;
+			best_run_index = i;
+		}
+
+		/*
+		 * Due to high overhead of gethrtime(), the performance data
+		 * here is inaccurate and much slower than it could be.
+		 * It's fine for our use though because only relative speed
```

Comment:
From what I've seen the performance difference between **scalar** vs **avx2** is so large reducing this inaccuracy wouldn't matter.  However, it's entirely possible we'll add some alternate hardware accelerated implementation were the difference isn't as large.  So I think it would be worth the effort to refactor the above loop as you suggested.